### ISSUE TITLE
Modify ajaxConvert to return a promise/thenable.

### ISSUE THREAD

**Issue by nathanhammond** [jquery/jquery#2532]
The internal `ajaxConvert` function always returns synchronously. This makes it impossible to support new techniques such as [async `JSON.parse`](http://azimi.me/2015/07/30/non-blocking-async-json-parse.html) by defining custom converters.

Considering:
- Ajax is already asynchronous.
- `ajaxConvert` is in an unreachable private scope.

I propose that we modify `ajaxConvert` to return a promise/thenable. If this is considered beneficial, I'll have a patch up with tests this week.

**Comment by dcherman** [comment]
Ajax is not always asynchronous; the `async: false` option still exists.

**Comment by nathanhammond** [comment]
@dcherman Yes, but the interface into it via `$.ajax` is already async (callbacks or thenable). I suppose that you could do something which created side effects in a synchronous manner without using callbacks, but then I believe you've explicitly opted not to use the `response`.

My goal is to be able to write async converters, and am able to make `ajaxConvert` run both synchronously and asynchronously if necessary.

**Comment by dmethvin** [comment]
That's not quite true, the plumbing calls inside are implicitly async up to this point so for example a uesr could have a `success` callback (not a Promise) but that could still use a converter.

I'm not sure whether we can support async filters and converters, they may break assumptions. It certainly would break assumptions of sync ajax but those are exceptions so maybe there is a way around it.

**Comment by markelog** [comment]
I guess you wouldn't use `ajax` if you have fetch right? What is your use-case, what use-cases are there?

**Comment by nathanhammond** [comment]
As it stands `fetch` is still a relatively immature API–we can still do much more with the `XHR` API by comparison. I suspect that they will be used in parallel for some time until achieving some sort of feature parity, plus the latent cost of migration efforts. I suspect that jQuery and `$.ajax` will be an important part of migrating people *to* `fetch`, and adding support for async converters creates a bit of a technical bridge.

The use case I have is an app with dependence upon Ember Data, which in turn delegates to `$.ajax` to make requests. We have server responses that push our parse times over the 16ms threshold and we want to handle that parsing asynchronously using `return (new Response(string)).json();`.

Goals:
- We don't want to replace the battle-tested `$.ajax` with a hand-rolled version.
- We want an intermediate step in the `XHR` => `fetch` transition.
- We want to enable every application using Ember Data to immediately benefit from this technique. (Improving an entire ecosystem.)
- Enable support for future asynchronous APIs which will allow other converters to become async as well.

I see this as simply an early encounter with the numerous async DOM and network APIs to come.

**Comment by dmethvin** [comment]
I know  you're not alone in this need, just last month I got a question at jQuery India from someone who was trying to parse a 50MB JSON response. :sob:

**Comment by gibson042** [comment]
Maybe I'm missing the point, but why is it necessary to address this with an AJAX converter rather than [with jqXHR#then](http://jsfiddle.net/ybnzhv1j/)?
```js
jQuery.ajax( url, { …, dataType: "text" })
  .then( convertAsPromise )
  .then( function( converted ) { … } )
```

**Comment by dcherman** [comment]
@gibson042 I thought about that as well, but one very unfortunate problem with that is `dataType` is coupled with setting the `Accept` header.  If you do what you're suggesting, then the `Accept` header is always `text/plain`.

There are only a couple workarounds to that:

1.  Set the `Accept` header manually and don't depend on jQuery to do it
2.  Depend on implementation details in $.ajax and set `options.dataTypes = ['text']` in a `beforeSend` handler.  A prefilter wouldn't work since they get run before the headers are set.

**Comment by gibson042** [comment]
​3. `converters` override: http://jsfiddle.net/ybnzhv1j/1/

Although we're admittedly in "implementation details" land any way you slice it, because of how magical `jQuery.ajax` tries to be.

**Comment by markelog** [comment]
> Depend on implementation details in $.ajax and set options.dataTypes = ['text'] in a beforeSend handler. A prefilter wouldn't work since they get run before the headers are set.

Why in `beforeSend`? http://jsfiddle.net/7eL6b2ky/ 

> Although we're admittedly in "implementation details" land any way you slice it

How so? `converters` property is a part of public api.

`jQuery.ajax` is the most extendable part of all jQuery, there is pretty much nothing you can't do with it, i'm pretty sure there is option somewhere that can make you a coffee if look hard enough.

**Comment by jaubourg** [comment]
I like the idea in principle and, yes, converters are actually the place to handle this to support content-type detection. As said in the PR/PoC, the solution needs to support chaining converters. Code would be much simpler if all request where async (`$.when` could be used to great effect internally) but we still have to handle sync requests...

**Comment by dcherman** [comment]
@markelog Your jsfiddle messes up the `Accept` header like I described in my previous comment.  @gibson042 provided another workable alternative though - just blow away the json converter, then you can maintain the correct header and still do manual parsing.

Out of curiosity, I wonder if there are any pros to setting the responseType to `json` on the XHR and using the `response` property in browsers that support it.  I tested it before and it still seemed to freeze the UI during parsing, so it doesn't seem any more beneficial than just piping responseText through JSON.parse.

**Comment by markelog** [comment]
> Your jsfiddle messes up the Accept header like I described in my previous comment.

I didn't say it didn't, i initialy thought you meant you that user could only defer parsing in `beforeSend` method, but know i see your point.

> I tested it before and it still seemed to freeze the UI during parsing,

Besides that, last time i checked, you will have issues if you json is incorrect. Our parsing will give meaningful output, when browser will not.

> but we still have to handle sync requests...

About that, since now sync requests are deprecated by the spec and spec encourages browsers to warn user about it, which vendors do. We, on the other hand, already partly deprecated `async: false` since 1.8. Can we have discussion about removing it? 

@timmywil, @dmethvin?

**Comment by nathanhammond** [comment]
@all Thank you for your comments and review. :) It seems like the general mood in this thread is that async converters are something we wish to enable. Past the first bits of conversation we've been focusing on implementation and alternative approaches to accomplishing the goal as opposed to the value of enabling, implying that we're in rough consensus as to the value it provides.

As to how it's implemented, part of my goal with this is to do the "right" thing, which I believe to be enabling async converters. Over the next day or two I'll iterate on my solution to address known gaps, use browser-safe solutions, add tests (tests currently pass without working converter chaining so there are some holes in coverage), and get it into a polished state for landing.

If I've misread the room's mood, let me know!

**Comment by gibson042** [comment]
I'm generally opposed, because I see it as adding further complexity to an already cumbersome system (given my druthers, converters theirselves would be fully separate from transports, with `.then` et al. bridging the two). And as I demonstrated above, it's already possible to do what you want with the existing API (and notably, using it in a way that _works around_ automatic behavior).

That said, though, I could be convinced by a particularly good PR.

**Comment by nathanhammond** [comment]
See my latest update here: https://github.com/jquery/jquery/pull/2534#issuecomment-130472792

`jQuery.Deferred` has painted us into a corner that I'm unsure we can get out of.

**Comment by jaubourg** [comment]
You have to test if `then` is a function (see `when`'s [code](https://github.com/jquery/jquery/blob/master/src/deferred.js#L330)).

Nothing to do with the implementation of `Deferred` that's how all Promise libraries detect thenables (hence the name, "thenables").

**Comment by nathanhammond** [comment]
@jaubourg Yes, but the problem is that this is a valid converter which is duck-typed into being a thenable:

```js
function(response) {
  return { value: "asdf", then: function() {} };
}
```

We're unable to protect the user in this case, assume their intent, or provide any error messages. I much prefer defensive APIs to duck typing, especially in library code, and I can't figure out a way to accomplish that while using `jQuery.Deferred`. Having disallowed values by implementation detail (especially in a non-recoverable manner) can result in a lot of nasty surprises: https://github.com/emberjs/ember.js/issues/4792

jQuery is in enough places that it seems inevitable that this would bite somebody.

**Comment by jaubourg** [comment]
I think you're over-thinking this.

First of all, jQuery does not support changing `Object.prototype` so it's a non issue here (the library will break long before you reach the conversion logic).

Secondly, __every__ javascript library handles a thenable like this (even [RSVP.js](https://github.com/tildeio/rsvp.js/blob/2136b263a72719a15d908c343e809df9fe604659/lib/rsvp/-internal.js) the promise library used inside _Ember.js_). As of today, it is pretty much a coding error to create an object with a `then` method that is not Promise-standard compliant (we got enough slack regarding this ourselves to know).

That being said, if you decide not to go through with this, you'll make a happy @gibson042, I'm sure, but it would be a shame to do so based on design considerations that do no apply to jQuery in the sense this would be the only place in the whole library they are considered.

**Comment by gibson042** [comment]
What @jaubourg said. _Every_ `then` method is assumed to honor [Promises/A+](https://promisesaplus.com/) at this point, both inside and outside of jQuery. It's not cause to block this.

**Comment by dmethvin** [comment]
We discussed in our [meeting](http://meetings.jquery.org/2015-08-31-core/) and decided that it added additional complexity inside the AJAX implementation for very little gain, since it can already be done with the current API. If this was added we'd need additional documentation, and our `$.ajax` docs are already the most complex API inside jQuery.

**Comment by nathanhammond** [comment]
I'm game to write the documentation as well if this ever becomes something we want to address in the future. I'd argue that it does add a lot of magic now, but at some point I feel like the entire `$.ajax` API should be async only, cleaning up the logic tremendously. :)

We'll go with a hack on our side to implement this. Thanks for the consideration!

**Comment by markelog** [comment]
I think we really should do something about ajax interface and approach to it in more systematic way

**Comment by dmethvin** [comment]
I think `$.ajax()` is what it is; trying to change its behavior now is bound to break and/or complicate it. We talked about creating a cleaner API based around `fetch()` but that API would need shimming and also `fetch()` currently lacks important features like `abort()` to stop a request.

**Comment by pensierinmusica** [comment]
@dmethvin in https://github.com/jquery/jquery/issues/2532#issuecomment-138980727 you wrote:
> it can already be done with the current API.

Probably I'm missing something... would you be so kind to give me a code example of how #2600 can be done with the current API?

Thanks!
